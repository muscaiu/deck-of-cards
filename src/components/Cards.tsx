import React, { useState } from "react";

import Card from "./Card";
import useFetchCards from "../hooks/useFetchCards";
import { CardInterface } from "../types/types";

import "./card.scss";

const Cards: React.FC = () => {
  const [{ cards, setCards, isLoading, isError }] = useFetchCards();
  const [selectedCards, setSelectedCards] = useState<CardInterface[]>([]);

  const handleCardClick = (card: CardInterface) => {
    const remainingCards = cards.filter((c) => card.value !== c.value);
    setCards(remainingCards);
    setSelectedCards([...selectedCards, card]);
  };

  const handleShuffle = () => {
    const shuffledCards = cards
      .sort(() => Math.random() - 0.5)
      .map((item) => item);

    setCards(shuffledCards);
  };

  const handleSortLowToHigh = () => {
    const sortedCards = selectedCards
      .sort((a, b) => a.sortOrder - b.sortOrder || a.value - b.value)
      .map((item) => item);

    setSelectedCards(sortedCards);
  };
  const handleSortHighToLow = () => {
    const sortedCards = selectedCards
      .sort((a, b) => a.sortOrder - b.sortOrder || b.value - a.value)
      .map((item) => item);

    setSelectedCards(sortedCards);
  };

  return (
    <>
      {isError && <div>Something went wrong...</div>}

      {isLoading ? (
        <div>Loading...</div>
      ) : (
        <div>
          <div className="sort-btn-container">
            <div className="shuffle-btn" onClick={handleShuffle}>
              shuffle
            </div>
          </div>

          <div className="cards">
            {cards &&
              cards.map((card, i) => (
                <Card key={i} card={card} onClick={handleCardClick} />
              ))}
          </div>

          <div className="sort-btn-container">
            <div className="sort-btn" onClick={handleSortHighToLow}>
              sort ↓
            </div>
            <div className="sort-btn" onClick={handleSortLowToHigh}>
              sort ↑
            </div>
          </div>

          <div className="selected-cards">
            {selectedCards &&
              selectedCards.map((card, i) => <Card key={i} card={card} />)}
          </div>
        </div>
      )}
    </>
  );
};

export default Cards;
