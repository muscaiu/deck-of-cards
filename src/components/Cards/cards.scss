.cards {
  transition: var(--transition);
  animation: fadeIn ease 2s;
}

.btn-container {
  display: flex;
  justify-content: center;

  .sort-btn:nth-child(2) {
    margin-left: 20px;
  }
}

.shuffle-btn,
.sort-btn {
  width: 270px;
  background: white;
  color: black;
  margin: 20px 0;
  font-size: 1.6rem;
  padding: 12px 36px;
  text-align: center;
  cursor: pointer;
  font-weight: 400;
  font-size: 21px;
  transition: var(--transition);
  animation: fadeIn ease 2s;

  &:hover {
    background-color: #b3b3b3;
  }
}

.info {
  text-align: center;
  color: #fff;
  margin-bottom: 20px;
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
