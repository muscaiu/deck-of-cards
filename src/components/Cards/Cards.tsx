import React, { useState } from "react";

import Card from "../Card/Card";
import useFetchCards from "../../hooks/useFetchCards";
import { CardInterface } from "../../types/types";

import "./cards.scss";

const Cards: React.FC = () => {
  const [{ cards, setCards, isLoading, isError }] = useFetchCards();
  const [selectedCards, setSelectedCards] = useState<CardInterface[]>([]);

  const handleCardClick = (card: CardInterface) => {
    const remainingCards = cards.filter((c) => card.value !== c.value);
    setCards(remainingCards);
    setSelectedCards([...selectedCards, card]);
  };

  const handleShuffle = () => {
    const shuffledCards = cards
      .sort(() => Math.random() - 0.5)
      .map((item) => item);

    setCards(shuffledCards);
  };

  const handleSortLowToHigh = () => {
    const sortedCards = selectedCards
      .sort((a, b) => a.sortOrder - b.sortOrder || a.value - b.value)
      .map((item) => item);

    setSelectedCards(sortedCards);
  };

  const handleSortHighToLow = () => {
    const sortedCards = selectedCards
      .sort((a, b) => a.sortOrder - b.sortOrder || b.value - a.value)
      .map((item) => item);

    setSelectedCards(sortedCards);
  };

  return (
    <>
      {isError && <div>Something went wrong...</div>}

      {isLoading ? (
        <div>Loading...</div>
      ) : (
        <div className="cards">
          <div className="btn-container">
            <div className="shuffle-btn" onClick={handleShuffle}>
              shuffle
            </div>
          </div>
          <p className="info">or click some cards</p>

          <div>
            {cards &&
              cards.map((card, i) => (
                <Card key={i} card={card} onClick={handleCardClick} />
              ))}
          </div>

          {selectedCards.length ? (
            <>
              <div className="btn-container">
                <div className="sort-btn" onClick={handleSortHighToLow}>
                  sort ↓
                </div>
                <div className="sort-btn" onClick={handleSortLowToHigh}>
                  sort ↑
                </div>
              </div>
              <div className="info">
                <p>you can now sort the selected cards</p>
                <p>
                  (order is: Clubs, Spades, Hearts, Diamonds; then by value: Ace
                  is high)
                </p>
              </div>
            </>
          ) : null}

          <div className="selected-cards">
            {selectedCards &&
              selectedCards.map((card, i) => <Card key={i} card={card} />)}
          </div>
        </div>
      )}
    </>
  );
};

export default Cards;
