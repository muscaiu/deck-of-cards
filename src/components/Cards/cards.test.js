import React from "react";
import { render } from "@testing-library/react";

import Cards from "./Cards";

const cards = [
  {
    img: "https://deck.of.cards/img/standard-deck/front-0.png",
    sortOrder: 1,
    value: 2,
  },
  {
    img: "https://deck.of.cards/img/standard-deck/front-1.png",
    sortOrder: 1,
    value: 3,
  },
];

describe("cards", () => {
  it("renders correctly", () => {
    const view = render(<Cards />);
    expect(view).toMatchSnapshot();
  });

  it("should render the shuffle button", () => {
    const { container } = render(<Cards />);
    expect(container.getElementsByClassName("shuffle-btn").length).toBe(1);
  });
});
