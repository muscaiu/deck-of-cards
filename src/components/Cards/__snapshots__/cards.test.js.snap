// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cards renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        Something went wrong...
      </div>
      <div
        class="cards"
      >
        <div
          class="btn-container"
        >
          <div
            class="shuffle-btn"
          >
            shuffle
          </div>
        </div>
        <p
          class="info"
        >
          or click some cards
        </p>
        <div />
        <div
          class="selected-cards"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      Something went wrong...
    </div>
    <div
      class="cards"
    >
      <div
        class="btn-container"
      >
        <div
          class="shuffle-btn"
        >
          shuffle
        </div>
      </div>
      <p
        class="info"
      >
        or click some cards
      </p>
      <div />
      <div
        class="selected-cards"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
